core/packages/contracts/src/InboundChannel.sol

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+import "openzeppelin/utils/cryptography/MerkleProof.sol";
+import "openzeppelin/access/AccessControl.sol";
+import "./ParachainClient.sol";
+import "./IRecipient.sol";
+import "./IVault.sol";
+
+contract InboundChannel is AccessControl {
+    // Nonce for each origin
+    mapping(bytes => uint64) public nonce;
+
+    // Registered message handlers
+    mapping(uint16 => Handler) handlers;
+
+    // Light client message verifier
+    IParachainClient public parachainClient;
+
+    // Relayers are rewarded from this vault
+    IVault public vault;
+
+    // The relayer reward for submitting a message
+    uint256 public reward;
+
+    // The governance contract, which is a proxy for Polkadot governance, administers via this role
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+
+    // Relayers must provide enough gas to cover message dispatch plus this buffer
+    uint256 public constant GAS_BUFFER = 24000;
+
+    // Inbound message from BridgeHub parachain
+    struct Message {
+        bytes origin;
+        uint64 nonce;
+        uint16 handler;
+        bytes payload;
+    }
+
+    // Message handler entry
+    struct Handler {
+        // Address of message handler which implements IRecipient
+        address recipient;
```

Comment:
should this be an IRecipient type or address type? i guess not a big deal either way?